### PR TITLE
Http session in options

### PR DESCRIPTION
**Pull request by rain0r**
Moved the menu item `Enable Session Tracking (Cookie)` from `Edit` to `Tools -> Options -> Connection`.

This way it can be persisted.

See https://github.com/zaproxy/zaproxy/issues/3394

### CODE REVIEW

File: src/org/parosproxy/paros/extension/option/OptionsCertificatePanel.java [hunk 1]
```diff
@@ -59,7 +59,6 @@
 import org.parosproxy.paros.model.Model;
 import org.parosproxy.paros.model.OptionsParam;
 import org.parosproxy.paros.view.AbstractParamPanel;
-import org.parosproxy.paros.view.View;
```
Comment (thc202) [inline src/org/parosproxy/paros/extension/option/OptionsCertificatePanel.java]: It seems the branch needs to be rebased, this was already done in a previous commit.

File: src/lang/Messages.properties [hunk 2]
```diff
@@ -841,6 +841,7 @@ conn.options.proxy.excluded.domain.dialog.remove.button.confirm = Remove
 conn.options.proxy.excluded.domain.dialog.remove.checkbox.label = Do not show this message again
 conn.options.proxy.useProxyChain = Use proxy chain
 conn.options.singleCookieRequestHeader = Single Cookie Request Header
+conn.options.httpStateEnabled = Enable HTTP state
```
Comment (thc202) [inline src/lang/Messages.properties]: Enable (Global) HTTP State?

File: src/org/parosproxy/paros/extension/option/OptionsConnectionPanel.java [hunk 3]
```diff
@@ -854,7 +858,19 @@ private JCheckBox getCheckBoxSingleCookeRequestHeader() {
         return checkBoxSingleCookieRequestHeader;
     }
 
-    private ProxyExcludedDomainsMultipleOptionsPanel getProxyExcludedDomainsPanel() {
```
Comment (thc202) [inline src/org/parosproxy/paros/extension/option/OptionsConnectionPanel.java]: This indentation change should be reverted (same for above unnecessary changes).

File: src/org/parosproxy/paros/extension/option/OptionsConnectionPanel.java [hunk 4]
```diff
@@ -854,7 +858,19 @@ private JCheckBox getCheckBoxSingleCookeRequestHeader() {
...
+    public JCheckBox getCheckBoxHttpStateEnabled() {
+        if (checkBoxHttpStateEnabled == null) {
+            checkBoxHttpStateEnabled =
+                    new JCheckBox(Constant.messages.getString("conn.options.httpStateEnabled"));
+        }
+        return checkBoxHttpStateEnabled;
+    }
+
+	public void setCheckBoxHttpStateEnabled(JCheckBox checkBoxHttpStateEnabled) {
```
Comment (thc202) [inline src/org/parosproxy/paros/extension/option/OptionsConnectionPanel.java]: Is this method really needed? It's not used (nor the use case very clear).

File: src/org/parosproxy/paros/network/ConnectionParam.java [hunk 5]
```diff
@@ -223,6 +225,12 @@ protected void parse() {
         }
         
         try {
+            this.httpStateEnabled = getConfig().getBoolean(HTTP_STATE_ENABLED, false);
+        } catch (ConversionException e) {
+            log.error("Error while loading the option httpStateEnabled: " + e.getMessage(), e);
+        }
+
```
Comment (thc202) [inline src/org/parosproxy/paros/network/ConnectionParam.java]: The extra spaces could be removed.

**Comment by thc202** [comment]
When do you plan to remove the menu item (if you still plan to)? Right now both options can be out of sync, enabled in one place and disabled in the other.

**Comment by thc202** [comment]
I should add that we can persist the state of the menu item without adding the new option to the options panel (which would probably be better at this stage?).

File: src/lang/Messages.properties (hunk 2 above)
Comment (rain0r) [inline src/lang/Messages.properties]: Will add the `(Global)`.

File: src/org/parosproxy/paros/extension/option/OptionsConnectionPanel.java (hunk 4 above)
Comment (thc202) [inline src/org/parosproxy/paros/extension/option/OptionsConnectionPanel.java]: I know it's a setter but do we really need it? What's the use case? If a new check box is set the UI will stop working as expected (the one previously added to the UI is not being replaced).

File: src/org/parosproxy/paros/extension/state/ExtensionState.java [hunk 6]
```diff
@@ -74,39 +71,10 @@ public void hook(ExtensionHook extensionHook) {
...
-		    menuSessionTrackingEnable.setText(Constant.messages.getString("menu.edit.enableTracking"));	// ZAP: i18n
-			getMenuResetSessionState().setEnabled(menuSessionTrackingEnable.isSelected());
-
-			menuSessionTrackingEnable.addItemListener(new java.awt.event.ItemListener() { 
-
-				@Override
-				public void itemStateChanged(java.awt.event.ItemEvent e) {    
-
-					getModel().getOptionsParam().getConnectionParam().setHttpStateEnabled(menuSessionTrackingEnable.isEnabled());
-					getMenuResetSessionState().setEnabled(menuSessionTrackingEnable.isSelected());
```
Comment (thc202) [inline src/org/parosproxy/paros/extension/state/ExtensionState.java]: We need more changes, the reset menu item will stop working as is (it's not being enabled/disabled based on the state of the new option).

File: src/org/parosproxy/paros/extension/option/OptionsConnectionPanel.java [hunk 7]
```diff
@@ -846,28 +850,44 @@ private ZapTextField getTxtTimeoutInSecs() {
 		}
 		return txtTimeoutInSecs;
 	}
-	
-    private JCheckBox getCheckBoxSingleCookeRequestHeader() {
```
Comment (thc202) [inline src/org/parosproxy/paros/extension/option/OptionsConnectionPanel.java]: Can we leave these formatting changes out?  (Same comment for ConnectionParam.)

File: src/org/parosproxy/paros/extension/state/ExtensionState.java [hunk 8]
```diff
@@ -74,39 +71,10 @@ public void hook(ExtensionHook extensionHook) {
...
-	
-	/**
-	 * This method initializes menuViewImage	
-	 * 	
-	 * @return javax.swing.JCheckBoxMenuItem	
-	 */    
-	private JCheckBoxMenuItem getMenuSessionTrackingEnable() {
-		if (menuSessionTrackingEnable == null) {
-		    menuSessionTrackingEnable = new JCheckBoxMenuItem();
-		    menuSessionTrackingEnable.setText(Constant.messages.getString("menu.edit.enableTracking"));	// ZAP: i18n
```
Comment (thc202) [inline src/org/parosproxy/paros/extension/state/ExtensionState.java]: The resource message should be removed (if not used in other places).

File: src/org/parosproxy/paros/extension/state/ExtensionState.java [hunk 9]
```diff
@@ -74,70 +66,14 @@ public void hook(ExtensionHook extensionHook) {
...
-
-          //  @jve:decl-index=0:}  //  @jve:decl-index=0:
-	/**
-	 * This method initializes jMenuItem	
-	 * 	
-	 * @return javax.swing.JMenuItem	
-	 */    
-	private ZapMenuItem getMenuResetSessionState() {
-		if (menuResetSessionState == null) {
-			menuResetSessionState = new ZapMenuItem("menu.edit.resetState");
```
Comment (thc202) [inline src/org/parosproxy/paros/extension/state/ExtensionState.java]: These (this and the one in the action) messages should be removed, if no longer in use (which don't seem to be).
Removing this option hampers the usability a bit though (but I guess it's fine since the state should not be reset that often).

**Comment by thc202** [comment]
There are some ZAP comments missing.

File: src/org/parosproxy/paros/extension/option/OptionsConnectionPanel.java [hunk 10]
```diff
@@ -96,6 +97,8 @@
 	private ZapTextField txtTimeoutInSecs = null;
 	private JPanel panelGeneral = null;
     private JCheckBox checkBoxSingleCookieRequestHeader;
+    // ZAP: Checkbox to Enable / Disable HTTP State
```
Comment (thc202) [inline src/org/parosproxy/paros/extension/option/OptionsConnectionPanel.java]: This inline ZAP comment can be removed, it's enough with the one added earlier.

File: src/org/parosproxy/paros/network/ConnectionParam.java [hunk 11]
```diff
@@ -86,6 +86,8 @@
 	private static final String PROXY_CHAIN_PROMPT = CONNECTION_BASE_KEY + ".proxyChain.prompt";
 	private static final String TIMEOUT_IN_SECS = CONNECTION_BASE_KEY + ".timeoutInSecs";
 	private static final String SINGLE_COOKIE_REQUEST_HEADER = CONNECTION_BASE_KEY + ".singleCookieRequestHeader";
+	// ZAP: Added option key to enable / disable HTTP State
```
Comment (thc202) [inline src/org/parosproxy/paros/network/ConnectionParam.java]: Better add the ZAP comment at the top of the class (inline comments are no longer added).

File: src/lang/Messages.properties [hunk 12]
```diff
@@ -841,6 +841,7 @@ conn.options.proxy.excluded.domain.dialog.remove.button.confirm = Remove
 conn.options.proxy.excluded.domain.dialog.remove.checkbox.label = Do not show this message again
 conn.options.proxy.useProxyChain = Use proxy chain
 conn.options.singleCookieRequestHeader = Single Cookie Request Header
+conn.options.httpStateEnabled = Enable (Global) HTTP state
```
Comment (kingthorin) [inline src/lang/Messages.properties]: Please add a leading cap to `state` so that it's in title caps.
https://github.com/zaproxy/zaproxy/wiki/DevGuidelines#text-titles-menu-items-options-and-tool-tips

File: src/org/parosproxy/paros/extension/option/OptionsConnectionPanel.java [hunk 13]
```diff
@@ -818,56 +822,73 @@ private JPanel getPanelGeneral() {
 			gbc.fill = java.awt.GridBagConstraints.HORIZONTAL;
 			gbc.insets = new java.awt.Insets(2,2,2,2);
 			gbc.anchor = java.awt.GridBagConstraints.WEST;
-            
+
```
Comment (kingthorin) [inline src/org/parosproxy/paros/extension/option/OptionsConnectionPanel.java]: This white space change should be reverted.

File: src/org/parosproxy/paros/extension/option/OptionsConnectionPanel.java [hunk 14]
```diff
@@ -818,56 +822,73 @@ private JPanel getPanelGeneral() {
...
+			panelGeneral.add(uaLabel, LayoutHelper.getGBC(0, 1, 1, 0.5D));
+			panelGeneral.add(getCommonUserAgents(), LayoutHelper.getGBC(1, 1, 1, 0.5D, new Insets(2, 2, 2, 2)));
+			panelGeneral.add(getDefaultUserAgent(), LayoutHelper.getGBC(0, 2, 2, 1.0D, new Insets(2, 2, 2, 2)));
 
-            panelGeneral.add(getCheckBoxSingleCookeRequestHeader(), gbc);
+			panelGeneral.add(getCheckBoxSingleCookeRequestHeader(), gbc);
+			panelGeneral.add(getCheckBoxHttpStateEnabled(), LayoutHelper.getGBC(0, 4, 3, 1.0D, 0, GridBagConstraints.HORIZONTAL, new Insets(16, 2, 2, 2)));
 
-}
+		}
```
Comment (kingthorin) [inline src/org/parosproxy/paros/extension/option/OptionsConnectionPanel.java]: Same

**Comment by thc202** [comment]
Thanks!

**Comment by kingthorin** [comment]
Thanks @rain0r!
